### PR TITLE
Arrow v10 + sparse union to represent OTLP any value

### PR DESCRIPTION
**Pull request by lquerel**
Sorry, this PR is huge. The migration to Arrow v10 brings new capabilities that have been leveraged to better represent OTLP AnyValue.

Arrow v10 supports dense and sparse union data types which allows us to represent all attributes (and log body) in a generic way. The major advantage is that we can now "statically" define Arrow schemas for traces, logs, and soon for the metrics. 

The previous approach was based on an intermediate representation allowing to dynamically determine the schema of each Arrow record according to the OTLP data streams to be encoded while optimizing the compression ratio. The use of unions makes it possible to eliminate this intermediate phase and to optimize CPU and memory performance (the encoding is 3 to 5 times faster).

This PR also improve the test coverage (~67%).

The AIR package will be removed once the metrics will be based on the new union data type.

### CODE REVIEW

File: pkg/otel/common/arrow/any_value.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+// appendF64 appends a new double value to the builder.
+func (b *AnyValueBuilder) appendF64(v float64) error {
+	b.builder.Append(F64Code)
+	b.f64Builder.Append(v)
+
+	b.strBuilder.AppendNull()
+	b.i64Builder.AppendNull()
+	b.boolBuilder.AppendNull()
+	b.binaryBuilder.AppendNull()
```
Comment (jmacd) [inline pkg/otel/common/arrow/any_value.go:167]: This _looks_ expensive (but doesn't mean it is). I was expecting to see a kind of union that appends a single value and that the underlying sparseness would be somehow implied by a single Append().

**Review by jmacd — APPROVED** [review]
Looks good. I am fairly sure I understand this change despite its size.

File: pkg/otel/common/arrow/any_value.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+	case pcommon.ValueTypeInt:
+		err = b.appendI64(av.Int())
+	case pcommon.ValueTypeDouble:
+		err = b.appendF64(av.Double())
+	case pcommon.ValueTypeBool:
+		err = b.appendBool(av.Bool())
+	case pcommon.ValueTypeBytes:
+		err = b.appendBinary(av.Bytes().AsRaw())
+	case pcommon.ValueTypeSlice:
+		// TODO implement this with CBOR encoding
```
[the remark this answers is not in the document]
Comment (lquerel) [reply pkg/otel/common/arrow/any_value.go:103]: Yes indeed.

File: pkg/arrow/utils.go [hunk 3]
```diff
@@ -122,13 +122,81 @@ func DataTypeToId(dt arrow.DataType) string {
...
+		id += "DenseUnion<>"
+	case *arrow.SparseUnionType:
+		// TODO implement
+		id += "SparseUnion<>"
+	case *arrow.MapType:
+		// TODO implement
+		id += "Map<>"
+	case *arrow.FixedSizeBinaryType:
+		// TODO implement
+		id += "FixedSizeBinary<>"
```
[the remark this answers is not in the document]
Comment (lquerel) [reply pkg/arrow/utils.go:136]: As the Arrow schemas for Traces and Logs is entirely static in this PR that should be ok. I will change that a little bit in a future PR to support some extreme cases (i.e. when an attribute cardinality is higher than 2^16).

File: pkg/arrow/utils.go [hunk 4]
```diff
@@ -190,12 +258,11 @@ type ListOfStructs struct {
 	end   int
 }
 
-// ListOfStructsFromRecord returns the struct type and an array of structs for a given field name.
-func ListOfStructsFromRecord(record arrow.Record, field string, row int) (*ListOfStructs, error) {
-	arr, err := Array(record, field)
-	if err != nil {
-		return nil, err
-	}
+// TODO remove bis once the other implementation is no longer used
```
[the remark this answers is not in the document]
Comment (lquerel) [reply pkg/arrow/utils.go:261]: I need to migrate the metrics with the union representation to get rid of this.

File: pkg/otel/common/arrow/any_value.go (hunk 1 above)
Comment (lquerel) [reply pkg/otel/common/arrow/any_value.go:167]: I will double check.

**Pull request merged**